pkg/controller/operators/adoption_controller.go

```diff
@@ -0,0 +1,438 @@
+	// Add watched types to scheme.
+	if err := AddToScheme(scheme); err != nil {
+		return nil, err
+	}
+
+	factory, err := decorators.NewSchemedOperatorFactory(scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AdoptionReconciler{
+		Client: cli,
+
+		log:     log,
+		factory: factory,
+	}, nil
+}
+
+var fieldOwner = client.FieldOwner("olm")
+
+// ReconcileSubscription labels the CSVs installed by a Subscription as components of an operator named after the subscribed package and install namespace.
+func (r *AdoptionReconciler) ReconcileSubscription(req ctrl.Request) (reconcile.Result, error) {
+	// Set up a convenient log object so we don't have to type request over and over again
+	log := r.log.WithValues("request", req)
+	log.V(1).Info("reconciling subscription")
+
+	// Fetch the Subscription from the cache
+	ctx := context.TODO()
+	in := &operatorsv1alpha1.Subscription{}
+	if err := r.Get(ctx, req.NamespacedName, in); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Could not find Subscription")
+		} else {
+			log.Error(err, "Error finding Subscription")
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// OLM generated Operators are named after their packages and further qualified by the install namespace
+	if in.Spec == nil || in.Spec.Package == "" {
+		log.Info("subscription missing status, ignoring")
+		return reconcile.Result{}, nil
+	}
+
+	// Wrap with convenience decorator
+	operator, err := r.factory.NewPackageOperator(in.Spec.Package, in.GetNamespace())
+	if err != nil {
+		log.Error(err, "Could not wrap Operator with convenience decorator")
+		return reconcile.Result{}, nil
+	}
+
+	// Ensure the subscription is adopted
+	var errs []error
+	out := in.DeepCopy()
+	adopted, err := operator.AdoptComponent(out)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if adopted {
+		if err := r.Patch(ctx, out, client.MergeFrom(in)); err != nil {
+			log.Error(err, "Error adopting Subscription")
+			errs = append(errs, err)
+		}
+	}
+
+	// Find the Subscription's CSVs and apply the component label if necessary
+	adoptCSV := func(name string) error {
+		csv := &operatorsv1alpha1.ClusterServiceVersion{}
+		if err := r.Get(ctx, types.NamespacedName{Namespace: in.GetNamespace(), Name: name}, csv); err != nil {
+			if apierrors.IsNotFound(err) {
+				err = nil
+			}
+
+			return err
+		}
+		log.Info("found CSV")
+
+		candidate := csv.DeepCopy()
+		adopted, err := operator.AdoptComponent(csv)
+		if err != nil {
+			return err
+		}
+
+		if adopted {
+			// Only update the CSV if freshly adopted
+			if err := r.Patch(ctx, csv, client.MergeFrom(candidate)); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	if name := in.Status.InstalledCSV; name != "" {
+		if err := adoptCSV(name); err != nil {
+			log.Error(err, "Error adopting installed CSV")
+			errs = append(errs, err)
+		}
+	}
+
+	if name := in.Status.CurrentCSV; name != "" && name != in.Status.InstalledCSV {
```

Comment:
do we need to adopt the current csv at all? it may not exist.

Reply:
we can probably drop this. I had thought there was an edge case where we can have both, but maybe that's not possible.